File: all-nodes-distance-k-in-binary-tree.cpp
class Solution {
public:
    void populateParentMap(TreeNode* root, map<TreeNode*, TreeNode*>& parentMap) {
        if (!root) return;

        if (root->left) {
            parentMap[root->left] = root;
            populateParentMap(root->left, parentMap);
        }
        if (root->right) {
            parentMap[root->right] = root;
            populateParentMap(root->right, parentMap);
        }
    }

    vector<int> distanceK(TreeNode* root, TreeNode* target, int k) {
        map<TreeNode*, TreeNode*> parentMap;
        populateParentMap(root, parentMap); // Populate the parent map.

        queue<TreeNode*> q;
        map<TreeNode*, bool> visited;
        q.push(target);
        visited[target] = true;

        int currentDistance = 0;

        // Perform BFS to find nodes at distance k.
        while (!q.empty()) {
            int size = q.size();
            if (currentDistance == k) break; // Stop when the desired distance is reached.

            for (int i = 0; i < size; i++) {
                TreeNode* node = q.front();
                q.pop();

                // Explore neighbors: left, right, and parent.
                if (node->left && !visited[node->left]) {
                    q.push(node->left);
                    visited[node->left] = true;
                }
                if (node->right && !visited[node->right]) {
                    q.push(node->right);
                    visited[node->right] = true;
                }
                if (parentMap[node] && !visited[parentMap[node]]) {
                    q.push(parentMap[node]);
                    visited[parentMap[node]] = true;
                }
            }

            currentDistance++;
        }

        // Collect all nodes at distance k.
        vector<int> result;
        while (!q.empty()) {
            result.push_back(q.front()->val);
            q.pop();
        }

        return result;
    }
};

File: all-nodes-distance-k-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def populate_parent_map(self, root, parent_map):
        if not root:
            return
        
        if root.left:
            parent_map[root.left] = root
            self.populate_parent_map(root.left, parent_map)
        if root.right:
            parent_map[root.right] = root
            self.populate_parent_map(root.right, parent_map)

    def distanceK(self, root: TreeNode, target: TreeNode, k: int):
        parent_map = {}
        self.populate_parent_map(root, parent_map)  # Populate the parent map.

        q = deque([target])
        visited = {target: True}
        current_distance = 0

        # Perform BFS to find nodes at distance k.
        while q:
            if current_distance == k:
                break  # Stop when the desired distance is reached.

            for i in range(0,len(q),1):
                node = q.popleft()
                
                # Explore neighbors: left, right, and parent.
                if node.left and node.left not in visited:
                    q.append(node.left)
                    visited[node.left] = True

                if node.right and node.right not in visited:
                    q.append(node.right)
                    visited[node.right] = True

                if parent_map.get(node) and parent_map[node] not in visited:
                    q.append(parent_map[node])
                    visited[parent_map[node]] = True
                    


            
            current_distance += 1

        # Collect all nodes at distance k.
        return [node.val for node in q]
